clang/lib/Driver/ToolChains/Clang.cpp

```diff
@@ -4154,7 +4154,9 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
       CmdArgs.push_back("-fenable-sycl-dae");
 
     // Pass the triple of host when doing SYCL
-    auto AuxT = llvm::Triple(llvm::sys::getProcessTriple());
+    llvm::Triple AuxT = C.getDefaultToolChain().getTriple();
+    if (Args.hasFlag(options::OPT_fsycl_device_only, OptSpecifier(), false))
+      AuxT = llvm::Triple(llvm::sys::getProcessTriple());
```

Comment:
Why not also do this for -fsycl-device-only?

Reply:
When compiling in -fsycl-device-only the default toolchain of the compilation is the device target not the host target. i don't think the host target is actually known when building in -fsycl-device-only.

Reply:
@mdtoguchi interesting question... But I have the feeling that compiling for the device is already a kind of cross-compilation per se.
Our use case is when we compile SYCL code on x86 for ACAP devices where the host is ARM64 and the devices other specific VLIW DSP.

Reply:
> When compiling in -fsycl-device-only the default toolchain of the compilation is the device target not the host target. i don't think the host target is actually known when building in -fsycl-device-only.

It's known (via `-aux-triple`) and it's required to fix the sizes of data types like `int` for both: host and device (and for some other reasons). Otherwise host/device ABI might be incompatible.

Reply:
-aux-triple is a cc1 only flag so the driver can't really know.
i changed it to be cc1 or dirver.

Reply:
> -aux-triple is a cc1 only flag so the driver can't really know.
> i changed it to be cc1 or dirver.

Sorry, if I confused you somehow, I don't quite get the reason to make `-aux-triple` a driver option.
Driver sets front-end options (i.e. cc1), so it "knowns" all cc1 options. I think what Mike is asking for is to configure `-aux-triple` value when `-fsycl-device-only` is set as well. There should be no reason to expose it in the driver. Right?

Reply:
> what i meant by "knows" is "is valid to provide to" and the driver can't receive a -aux-triple option since they are cc1 only.

With this definition of "knows", the driver should not "know" "aux-triple".

> currently the driver can add a -aux-triple flag to a cc1 invocation
> but the driver can't be specified a host target because -fsycl-device-only is considered a cross-compilation so all cross-compilation flags are used for the device target. and -aux-triple isn't a valid driver option.

`-fsycl-device-only` says the compiler to produce device part of the final binary, which can be executed on particular system. "system" here is a combination of a host+device. Compiling the same program with `-fsycl-device-only` on ARM+NVIDIA GPU and INTEL+NVIDIA GPU will produce different results even if NVIDIA GPU is the same device.

> how is the driver supposed to know or be specified the host target except by guessing that the host target is the same device that is running the compilation (which is wrong when cross-compiling) ?

I expected this patch to adjust host target based on `-target` option value, rather than adding new `-aux-target`.

Reply:
> I expected this patch to adjust host target based on `-target` option value, rather than adding new `-aux-target`.
the issue is that -target is already used to specify the target device.

Reply:
Using SYCL terms:
- we use `-fsycl-targets=` to specify target "device"
- we can use `-target` to specify target "host"

AFAIK, `-target` is not used to specify offload device for other models like CUDA, HIP or OpenMP.

Reply:
> Using SYCL terms:
> 
>     * we use `-fsycl-targets=` to specify target "device"
> 
>     * we can use `-target` to specify target "host"
> 
the behavior you described is what i observed without the -fsycl-device-only and it would make a lot of sense if it was the same behavior with -fsycl-device-only but this isn't what i observed (on current tip of sycl).
here is what I see:
$ clang++ -fsycl -fsycl-device-only test.cpp -fsycl-targets=nvptx64-nvidia-cuda-sycldevice -###
...
clang-12: error: The option -fsycl-targets must be used in conjunction with -fsycl to enable offloading.
... clang-12" "-cc1" "-triple" "spir64-unknown-linux-sycldevice" ... "-aux-triple" "x86_64-unknown-linux-gnu" ...
$ clang++ -fsycl -fsycl-device-only test.cpp -target aarch64-linux -###
... clang-12" "-cc1" "-triple" "aarch64-unknown-linux" ... "-aux-triple" "x86_64-unknown-linux-gnu" ...

maybe this isn't the intended behavior but how are we supposed to specify the target device in -fsycl-device-only if i changed -target to specify the target host ?

Reply:
I suppose this might be related to https://github.com/intel/llvm/issues/1814. I'm not sure if we can address Mike's comment easily.
@mdtoguchi, should we resolve it in scope of #1814 or there is an easy fix here?

Reply:
Use of `-fsycl-device-only` should behave in a simlar fashion to `-fsycl` in regards to the device compilation step.  The disconnects need to be addressed as stated in #1814 .  The behavior being seen with the error emitted when using `-fsycl-targets` with `-fsycl -fsycl-device-only` is not intentional.

Reply:
@mdtoguchi, let me check that I understand current status correctly.
Am I right that handling `-fsycl-device-only`+`-target` together is blocked by #1814 and we agree to resolve it separately?

Anyway, I think we all agree that we should not add `-aux-triple` driver option. Right?

Reply:
> @mdtoguchi, let me check that I understand current status correctly.
> Am I right that handling `-fsycl-device-only`+`-target` together is blocked by #1814 and we agree to resolve it separately?
I brought back the patch before a driver -aux-triple to resolve it separately.

> Anyway, I think we all agree that we should not add `-aux-triple` driver option. Right?
I agree and I think the behavior you described is a better solution than a driver -aux-triple.

Reply:
Agree. #1814 should be resolved separately and we do not need `-aux-triple` driver option.